src/database-postgresql.cpp

```diff
@@ -0,0 +1,505 @@
+		} else if (m_conf.exists("pg_port") && !m_conf.exists("pg_host")) {
+		    m_port = m_conf.get("pg_port");
+		    m_host = std::string("localhost");
+		} else if (!m_conf.exists("pg_port") && m_conf.exists("pg_host")) {
+		    m_port = "5432";
+		} else {
+		    infostream << "Postgres: Using socket connection.\n";
+		}
+
+		// password
+		if (m_conf.exists("pg_password"))
+		    m_password = m_conf.get("pg_password");
+		else
+		    infostream << "Using no password.\n";
+
+		// database name
+		if (m_conf.exists("pg_dbname")) {
+		    m_dbname = m_conf.get("pg_dbname");
+		} else {
+		    m_dbname = std::string("minetest");
+		    infostream << "Postges: Using default database name, minetest.\n";
+		}
+
+	} catch(SettingNotFoundException e) {
+		throw SettingNotFoundException("Set database in world.mt to use the pgsql backend");
+	}
+
+	m_conninfo = "user=" + m_user + " dbname=" + m_dbname;
+
+	if (m_host.size() > 0)
+		m_conninfo = m_conninfo + " host=" + m_host;
+
+	if (m_port.size() > 0)
+		m_conninfo = m_conninfo + " port=" + m_port;
+
+	if (m_password.size() > 0)
+		m_conninfo = m_conninfo + " password=" + m_password;
+
+	infostream << "Postgres connection info: " << m_conninfo << std::endl;
+}
+
+bool Database_PostgreSQL::initialized(void)
+{
+	return PQstatus(m_conn) == CONNECTION_OK ? true : false;
+}
+
+void Database_PostgreSQL::beginSave()
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	m_res = PQexec(m_conn, "BEGIN;");
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		break;
+	default:
+		throw BaseException(PQresultErrorMessage(m_res));
+		break;
+	}
+
+	PQclear(m_res);
+	return;
+}
+
+void Database_PostgreSQL::endSave()
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	m_res = PQexec(m_conn, "COMMIT;");
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		break;
+	default:
+		throw BaseException(PQresultErrorMessage(m_res));
+		break;
+	}
+
+	PQclear(m_res);
+	return;
+}
+
+/**
+ * No need to do anything.
+ */
+void Database_PostgreSQL::createDirs(std::string path)
+{
+	return;
+}
+
+bool Database_PostgreSQL::deleteBlock(const v3s16 &pos)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[1];
+
+	param_values[0] = posTochar(pos);
+
+	m_res = PQexecPrepared(m_conn, "delete_block", 1, param_values, NULL, NULL, 1); 
+
+	switch(PQresultStatus(m_res)) {
+	case PGRES_COMMAND_OK:
+		PQclear(m_res);
+		return true;
+		break;
+	default:
+		errorstream << "Error Message:" << PQresultErrorMessage(m_res) << std::endl;
+		PQclear(m_res);
+		return false;
+		break;
+	}
+
+}
+
+bool Database_PostgreSQL::saveBlock(const v3s16 &pos, const std::string &data)
+{
+	DSTACK(__FUNCTION_NAME);
+
+	verifyDatabase();
+
+	const char* param_values[2];
+	int param_format[2];
+	int param_length[2];
+
+	std::string old_data = loadBlock(pos);
+
+	// This is of interest. I have noticed there are a large number of
+	// saveBlock calls where the new and old data are exactly the same.
```

Comment:
It should be investigated why this happens and the bug should be fixed at the source rather than having this workaround.